gk/main.c

```diff
@@ -2132,41 +2150,44 @@ add_ggu_policy_bulk(struct gk_add_policy **policies, int num_policies,
 
 	for (i = 0; i < num_policies; i++) {
 		int pos = pos_arr[i];
-		struct flow_entry *fe;
-		struct ggu_policy *policy = &policies[i]->policy;
 
-		if (pos >= 0)
-			fe = &instance->ip_flow_entry_table[pos];
-		else if (instance->num_scan_del > 0)
-			continue;
-		else {
-			struct gk_fib *fib = look_up_fib(
-				&gk_conf->lpm_tbl, &policy->flow);
-			if (fib == NULL || fib->action !=
-					GK_FWD_GRANTOR) {
-				/*
-				 * Drop this solicitation to add
-				 * a policy decision.
-				 */
-				char err_msg[128];
-				int ret = snprintf(err_msg, sizeof(err_msg),
-					"gk: at %s initialize flow entry error",
-					__func__);
-				RTE_VERIFY(ret > 0 &&
-					ret < (int)sizeof(err_msg));
-				print_flow_err_msg(
-					&policy->flow, err_msg);
-				continue;
-			}
+		if (pos >= 0) {
+			struct ggu_policy *policy =
+				&policies[i]->policy;
+			struct flow_entry *fe =
+				&instance->ip_flow_entry_table[pos];
 
-			fe = add_new_flow_from_policy(fib,
-				policy, instance, gk_conf,
-				flow_hash_val_arr[i]);
-			if (fe == NULL)
-				continue;
+			update_flow_entry(fe, policy);
+		} else if (flow_arr[i]->proto == RTE_ETHER_TYPE_IPV4) {
+			lpm_lookup_pos[num_lpm_lookups] = i;
+			flows[num_lpm_lookups] = flow_arr[i];
+			num_lpm_lookups++;
+		} else {
+			lpm6_lookup_pos[num_lpm6_lookups] = i;
+			flows6[num_lpm6_lookups] = flow_arr[i];
+			num_lpm6_lookups++;
 		}
+	}
+
+	if (instance->num_scan_del > 0)
+		return;
+
+	/* The remaining flows need LPM lookups. */
+	lookup_fib_bulk(&gk_conf->lpm_tbl, flows, num_lpm_lookups, fibs);
+	lookup_fib6_bulk(&gk_conf->lpm_tbl, flows6, num_lpm6_lookups, fibs6);
+
+	for (i = 0; i < num_lpm_lookups; i++) {
+		int fidx = lpm_lookup_pos[i];
+
+		update_flow_table(fibs[i], &policies[fidx]->policy,
+			instance, gk_conf, flow_hash_val_arr[fidx]);
```

Comment:
Replace `flow_hash_val_arr[fidx]` with `policies[fidx]->flow_hash_val`. This change seems just style, but notice that `policies[fidx]` has already been resolved in `&policies[fidx]->policy`, so getting the second field, namely `flow_hash_val`, is cheaper than resolving `flow_hash_val_arr[fidx]` from scratch.

Reply:
Good point!